### ISSUE TITLE
A question about theming / styling / etc..

### ISSUE THREAD

**Issue by JovanVeljanoski** [widgetti/solara#861]
Hi, 

A question about theming in solara.
I have a `assets/theme.js` in the root of my project, a file with which I hope to control the theme(s) of the application. It seems to be picked up correctly by solara, and works well. 

My question is, how can I modify the.. (I do not know what i the correct terminology here) the "true background" of the application, the "canvas" that goes black for example when one is using the pre-defined dark-mode (also the text letters go white in that case). Is that something that can be modified per app?

[A small pycafe example that hopefully shows what I am talking about](https://py.cafe/snippet/solara/v1?pycafe-app-view=false&pycafe-edit-enable=true#c=H4sIAHyINmcEA-1Y3W7jNhZ-FcFz4RSINZYs_wUQsLEnngK73Q0GKfaiKQpaoi2OJVJDUknUQd6936FkS04nmPS6lS9snnP4nf9Dyl8HiUr54GogilJp6xmVM83u5b38V_PTTxQ4kkt7L1O-827Znl_8cEUSHp5HYbN2k79WeVXIC5aLvYyHCbZwPSRREqTnTJjp9MLYOufx8FGkNrvypuNx-XS2gZ6zTY2GlyL0tBJ3_Mle3A9-5Hmu7gc__FmuD_dJPX4Ti55zvDu13-fcg2eZfZ8yffAKhO3qmxroaXfnbOvbjBdC7v07fPMG54JLts35b6yyKt6w3PBvwLQQq8paJS-AxPN4eKtFwXQ9vPQSlSsdD8uW8FaAFUsOe60qmXYY2472Vpg7XXGvt--vYR2DmnFvePfp5xtvdb3-98dP__v5vx-G3tYp8kymqjz1ttxDABt0T-3cYpgw-cAMlAqf-47UqcWSWc9WWhrPJUrI5psS5iNhg8uB5l8qoZEOaQ1Kv7HqXoJj65KaoaFgXdY2UxKUslapSPnoYeyHUz8AK2e1quzg6uvggWsjSCoEtFIWZUXkFktjdTlIMpGnmkPolxPHsq3hFkxX_4OrYDq-HKBW_t8sJ83qR05F1yxFim07kfMVUA3XayUtE5LrVzSQ6GjbyEKkZIQ7GDz_-nx5kjlZ8S0AMLt9rCz9sh6c7-3YyEPB_c_mVQFmoMi87-T6ZrwhTN8PTevq98JyjMiZASc1WYsazKdv0YkhBzLLv6f0KEda6fN86dKDAoSMZAXJtBHyeVHaeqPylOvbnCU8cz8JGEpQtRB1UT7fdgpsX-6h4lbsajd9jBd7X5uudJPs6riip50kV97w3U2wurmZo7_wvH_v_cSqB94JGg50tJQTXS9uZusliULwlmFGftSct4cDPSyhg4BEN5vNch01qCT9HzLB-wjXOumuj50di830hsBP0nsC70l1O02ld4jU2TY8b9jJtVba7dtMJ9Px8LJjCblTxInC6-lm2ueYCn4ZQ8zZbLWaXfeZj0xLIffE3GzW1-HiyHxuv2kgvRb82XWwWgKOyLD-FideJ3cW-8licTNZkyTkNkpzg3C-Gv3Vehx-aGBJ_qfKWBzBnWgXGhKProNoQdgQ_UDj84UdXbSjkD4NMITXGdOJYnkn24vv8po861jH-IbL1Wwz63N68V0E6_kCddMxz-K7mkdwq2FSfJ8xyHu98U9T_NMUr0f_b9UUv-LSopIDnTp0WCEGaIV3XsYMLt_peLYY83AZbKMw2LHtbJsGk0UwjqJFuJjOd_eSyVqoOI78yB9jpfdKhqNktxNxHE784AVxtBUyhT0G3MAP_bH3Dlew5EBCOH7jOPAnbovBde_AJckBOQDFWk2rFpNSkLA8j2NcvRyB494sdlVuVFVGZFAQ-iHoOWdJFsczf0xiCRbcKpUDawpdQE64xkaydwz0uR9iV-NA4Ae41WGFROEsHUmlC7zH_M51HE-wmQRzkRzieAGr5lipomgsAivl22pf1gQD9snRFHWpmVXAADjB4wWqMjx9Kpw3cyLxJ57AHbkn_53DO4Tks1HSwP6CETl0DlHeKH9kEayNQPmSSlLqTM-CgEADl52sKpiE-S46brdIJbAmtE2U9YFriXs8YoXbLDwAiS66Mbk3I7SWMNpzCesohm30wcD0dBFzLjVrhNW53-oqa1xp97iTNPGagmRUWiEYuCdT-GkzBD_zFMHHaulWQn5moXOPbKIYlEqgUF0SHHI_LvBs7k9Aq2Ar1yMkCCVNGuFVj4wZQAlAujsif4Coc2rWh8ClEJd5lwmoOyXynD3CT3eTc_tRWieBnG1HZb0vjtiuwHu8U1DImwD7MN4OqXpESFCO3XIkLHBOThO5Kg3bwQ-qEbfRlrmyudiOhMxx6yR1rjALhNlVl4thIYytiEtQWMttkuO2KBKXrJ6LxGjCh51OKyhKwl2QUKgzKhy53VFfgIKKGzsKxuuImVomNBoCRJN2Ksu3SqFQZwhPT0dLH5lMtM1DEFXRtE7odCio1HjTQpTmyBngSiZTlaD6EUVQ23IvmTZQOUbHIR4lfyp5AsMiH6V00liiZXHFRlO7AM2pEkuBvwZo_sBLB58zS26lgtAjGIU4lVoVHFeHypzKCuY28mCVdkST5SBAp8hGCHxpqFEoNEt_0VmA279WNLUbAwig0hwFyFyWQmd9nZA73FVeSM6digj5dv1Y1oWrDLw5Y1aii8CDB0viNG-mmbVlg-jE8UpqaMQTCe3ltLiOTpnljaUhYke-NnRqrVGO_ycaM2AssWpGs4pGqkP9vfgCJkZML8gaYxdv7JQZlwt6u0ZZOOMnrs31LplMJssRfBZQr6AAeJRt4iwXs5cc6NIiwSwPJsAkBKwoEzT9jmrK1LguwQbInMwxXKah1TjUTqKY6FXZHgUzGkjw2ogn4iO6gDdS4BhwFewOiub93y3dZGrWp_Hwgly5A-RIQqcKjBfyHhMKFIsDjKIOvHbc0CUj59a6opy4sYrappmSupKGP9iIQ6FOjMFABLiTUbqRcI0CF4UFCnxCQ0SES2-aFJM_5RmmoXrDaLyg6oIYzpuzUopQmuBUWmC6FdQSsK1Vi3mCOTMhHJfN6kFgpiLfsB0O3ctHZpMsVTjBzgvD0enUggLox-aOlbj_PNp6hemPHOMG__FAFOXlcg4SxznbXiJo0sJCEA0QcEQf-5Jy3Eg3DMiiahBE0Jp5ixvD0dXGRFJoDG5j5ksuLMdq8PwHxz7yjwkVAAA)

(in my app the theme is applied, here it is applied to the pycafe component - that was not my intention but funny side-effect). 

In general, is there a prefered way to set the true background to a user defined colour/picture/gradient etc..

In addition, regarding the `solara.Style(..)` component. 
Say we defined a CSS file that has bunch of classes.
1. Where in our component / page structure should we load the file (i.e. where should the `solara.Style(myccs.css)` so that the relevant components can have access to the defined classes in the .css file

2. Sometimes..(through a lot of trial and error) i found out that I need to do something like a .. "hard overwrite" of the vue classes to make the styling work, i.e. my css file has something like 

```css
.logo-button.v-btn {
    background: none !important;
    box-shadow: none !important;
    /*other definitions */
}
```

instead of 

```css
.logo-button {
    background: none !important;
    box-shadow: none !important;
    /*other definitions */
}
```
which was what I expected. (not, this example is in a custom Layout component..)

Then I apply the css like this in python

```python
solara.Style('path/to/mycss.css')
solara.Button(children=list_of_kids, on_click=my_callable, flat=True, classes=['logo-button'])
```

Is there some kind of best practice for this, or an expectation of how styling should be approached?

Many many thanks!

**Comment by JovanVeljanoski** [comment]
After some word, LLMs, and trial and error i think i have answered my own question, and I will write it here in case anyone else is interested in the same thing (but also so I can look it up in the future :P ).

_Warning: Most of what I am writing are assumptions and observations.. might be wrong or wrong way of thinking / approach. Follow the maintainers advice / suggestion if available_ .

If one wants a "total" control of everything styling, one should start with defining a custom layout. There, in the layout component, you can add CSS styling to define how the application background, etc.. will behave. 
Then, since solara is useing vue2/vuetifyjs under the hood, depending on how "deep" one wants to modify the components, sometimes it is necessary (or just easier) to modify the vue components directly, as they propagate throught the vue template, i.e.

```css
.v-application {
    background: #e8f0fe !important;
}
```

The above is an aggresive approach. 

One can also choose to a more fine-grained control, and style each individual component
```python
import solara
import reacton.ipyvuetify as v


v.App(class_="my_class")
v.AppBar(class_="my_class_app_bar")
```

but in this case, the style does not.. cascade so you really need to style everything, depending on what you need. Probably a combination of both is needed. 

I know nothing about CSS or best practices there, so doing stuff/discovering as I go along. 

On the topic of `solara.Style(..)` i found that only works if the argument is  of type `Path()`, and just string does not work for me (although the type hint suggests that it should?). 
I still do not know where this should be added.. but I just add it as the first thing in a component, and that seems to work generally ok. 


At the end of the day, it is nice to know that you can do whatever you want, and you can impose your style to solara/reacton components (if you know what you are doing). There are some steps to be done, but if one is familiar with CSS + vue, everything should be possible. 

To mods: Feel free to close this issue

**Comment by JovanVeljanoski** [comment]
Hi @iisakkirotko 

Thanks for the comprehensive answer - makes things way more clear! 
Also yeah, I think it would be quite handy at times to have a decorator as `@component_blank`. Anything to ease the friction between solara and vue(tify) would be much appreciated!

Thank you!

**Comment by JovanVeljanoski** [comment]
Just a quick update, in case anyone else is stuck on doing styling etc.. 
(and I am very much a noob on any of these topics..)

I realized that if your .css file imports other files for example

```css
@import './colors.css';
@import './fonts.css';

.my-classy-button{
    /* Colors only */
    background-color: var(--Primary-primary-100);
    color: var(--Text-text-main);
}
```

Then in solara one needs to apply all of the styles for the importing to work, i.e.

```python
@solara.component
def Page():
    with solara.Column(align='center'):
        solara.Style(value=Path(__file__).parent.parent / 'assets/styles.css')
        solara.Style(value=Path(__file__).parent.parent / 'assets/colors.css')
        solara.Style(value=Path(__file__).parent.parent / 'assets/fonts.css')
        solara.Button('Click me', classes=['my-classy-button'])
        solara.Button('Click me but unstyled')
```

Maybe this is obvious to the front-end experts, but to me it was not!

**Comment by maartenbreddels** [comment]
@JovanVeljanoski could you confirm that you see a 404 http request in the dev console for `http://localhost:8765/colors.css` ?

If so, we might have to not embed if we find a `@import` in https://github.com/widgetti/solara/blob/51cbfa970d42e5ff2c2b25de268e951677013467/solara/server/server.py#L321 (cc @iisakkirotko )

**Comment by JovanVeljanoski** [comment]
Hi, 

@maartenbreddels i am not sure what you mean.. `http://localhost:8765/colors.css` just says "Page not found by Solara router", which is expected.. There is nothing in the console log about it. 

On another note.. which might or might not be related.. is that when using `@solara.component_vue`, the styles "leak" from once component to another, which is kind of annoying, and probably a bug. 

Say you do something like

```vue
<template>
  <v-col>
    <div class="my-styling">My fancy text</div>
  </v-col>
</template>

<style scoped>
  .my-styling {
    font-weight: bold;
  }
</style>
```

And then say we have another component

```
<template>
  <v-col>
    <div class="my-styling">Another fancy text</div>
  </v-col>
</template>

<style scoped>
  .my-styling {
    font-weight: bold;
    color: red
  }
</style>
```

Then in the solara app, even if these two components are used in separate pages (same app tho), the css styling seems to be "global", i.e. the last one seems to overwrite all previous ones, even though they should be  "scoped",  i.e. only applied to the component in question (as far as my limited understanding goes).  

So in summary, you can't use css classes with the same name, even if they are "scoped" and part of different components that have nothing to do with each other but are being used in the same solara app. 

I hope this makes sense!
